urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/SevenZip/Compression/LZ/BinTree.java

```diff
@@ -0,0 +1,431 @@
+                    }
+                    distances[offset++] = maxLen = 3;
+                    distances[offset++] = _pos - curMatch3 - 1;
+                    curMatch2 = curMatch3;
+                }
+            }
+            if (offset != 0 && curMatch2 == curMatch)
+            {
+                offset -= 2;
+                maxLen = kStartMaxLen;
+            }
+        }
+        
+        _hash[kFixHashSize + hashValue] = _pos;
+        
+        int ptr0 = (_cyclicBufferPos << 1) + 1;
+        int ptr1 = _cyclicBufferPos << 1;
+        
+        int len0, len1;
+        len0 = len1 = kNumHashDirectBytes;
+        
+        if (kNumHashDirectBytes != 0)
+        {
+            if (curMatch > matchMinPos)
+            {
+                if (_bufferBase[_bufferOffset + curMatch + kNumHashDirectBytes] != _bufferBase[cur
+                        + kNumHashDirectBytes])
+                {
+                    distances[offset++] = maxLen = kNumHashDirectBytes;
+                    distances[offset++] = _pos - curMatch - 1;
+                }
+            }
+        }
+        
+        int count = _cutValue;
+        
+        while (true)
+        {
+            if (curMatch <= matchMinPos || count-- == 0)
+            {
+                _son[ptr0] = _son[ptr1] = kEmptyHashValue;
+                break;
+            }
+            final int delta = _pos - curMatch;
+            final int cyclicPos = (delta <= _cyclicBufferPos ? _cyclicBufferPos
+                    - delta : _cyclicBufferPos - delta + _cyclicBufferSize) << 1;
+            
+            final int pby1 = _bufferOffset + curMatch;
+            int len = Math.min(len0, len1);
+            if (_bufferBase[pby1 + len] == _bufferBase[cur + len])
+            {
+                while (++len != lenLimit)
+                {
+                    if (_bufferBase[pby1 + len] != _bufferBase[cur + len])
+                    {
+                        break;
+                    }
+                }
+                if (maxLen < len)
+                {
+                    distances[offset++] = maxLen = len;
+                    distances[offset++] = delta - 1;
+                    if (len == lenLimit)
+                    {
+                        _son[ptr1] = _son[cyclicPos];
+                        _son[ptr0] = _son[cyclicPos + 1];
+                        break;
+                    }
+                }
+            }
+            if ((_bufferBase[pby1 + len] & 0xFF) < (_bufferBase[cur + len] & 0xFF))
+            {
+                _son[ptr1] = curMatch;
+                ptr1 = cyclicPos + 1;
+                curMatch = _son[ptr1];
+                len1 = len;
+            }
+            else
+            {
+                _son[ptr0] = curMatch;
+                ptr0 = cyclicPos;
+                curMatch = _son[ptr0];
+                len0 = len;
+            }
+        }
+        MovePos();
+        return offset;
+    }
+    
+    public void Skip(int num) throws IOException
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.